arch/x86/core/ia32/x86_mmu.c

```diff
@@ -344,6 +360,169 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
 	}
 }
 
+static char __aligned(MMU_PAGE_SIZE)
+	page_pool[MMU_PAGE_SIZE * CONFIG_X86_MMU_PAGE_POOL_PAGES];
+
+static char *page_pos = page_pool + sizeof(page_pool);
+
+static void *get_page(void)
+{
+	page_pos -= MMU_PAGE_SIZE;
+
+	__ASSERT(page_pos >= page_pool, "out of MMU pages\n");
+
+	return page_pos;
+}
+
+__aligned(0x20) struct x86_mmu_pdpt z_x86_kernel_pdpt;
+#ifdef CONFIG_X86_KPTI
+__aligned(0x20) struct x86_mmu_pdpt z_x86_user_pdpt;
+#endif
+
+extern char z_shared_kernel_page_start[];
+
+static inline bool is_within_system_ram(uintptr_t addr)
+{
+	return (addr >= DT_PHYS_RAM_ADDR) &&
+		(addr < (DT_PHYS_RAM_ADDR + (DT_RAM_SIZE * 1024U)));
+}
+
+static void add_mmu_region_page(struct x86_mmu_pdpt *pdpt, uintptr_t addr,
+				u64_t flags, bool user_table)
+{
+	union x86_mmu_pdpte *pdpte;
+	struct x86_mmu_pd *pd;
+	union x86_mmu_pde_pt *pde;
+	struct x86_mmu_pt *pt;
+	union x86_mmu_pte *pte;
+
+#ifdef CONFIG_X86_KPTI
+	/* If we are generating a page table for user mode, and this address
+	 * does not have the user flag set, and this address falls outside
+	 * of system RAM, then don't bother generating any tables for it,
+	 * we will never need them later as memory domains are limited to
+	 * regions within system RAM.
+	 */
+	if (user_table && (flags & MMU_ENTRY_USER) == 0 &&
+	    !is_within_system_ram(addr)) {
+		return;
+	}
+#endif
+
+	/* Setup the PDPTE entry for the address, creating a page directory
+	 * if one didn't exist
+	 */
+	pdpte = &pdpt->entry[MMU_PDPTE_NUM(addr)];
+	if (pdpte->p == 0) {
+		pd = get_page();
+		pdpte->pd = ((uintptr_t)pd) >> MMU_PAGE_SHIFT;
+	} else {
+		pd = (struct x86_mmu_pd *)(pdpte->pd << MMU_PAGE_SHIFT);
+	}
+	pdpte->value |= (flags & PDPTE_FLAGS_MASK);
+
+	/* Setup the PDE entry for the address, creating a page table
+	 * if necessary
+	 */
+	pde = &pd->entry[MMU_PDE_NUM(addr)].pt;
+	if (pde->p == 0) {
+		pt = get_page();
+		pde->pt = ((uintptr_t)pt) >> MMU_PAGE_SHIFT;
+	} else {
+		pt = (struct x86_mmu_pt *)(pde->pt << MMU_PAGE_SHIFT);
+	}
+	pde->value |= (flags & PDE_FLAGS_MASK);
+
+	/* Execute disable bit needs special handling, we should only set it
+	 * at the page directory level if ALL pages have XD set (instead of
+	 * just one).
+	 *
+	 * Use the 'ignored2' field to store a marker on whether any
+	 * configured region allows execution, the CPU never looks at
+	 * or modifies it.
+	 */
+	if ((flags & MMU_ENTRY_EXECUTE_DISABLE) == 0) {
+		pde->ignored2 = 1;
```

Comment:
Style: given the subtlety, wouldn't be a bad idea to wrap the usage of ignored2 into an API like "set/get_page_partial_nx()" or something.